auth/src/main/scala/cromwell/pipeline/auth/AuthUtils.scala

```diff
@@ -5,57 +5,73 @@ import cromwell.pipeline.utils.AuthConfig
-      val newRefreshToken = getAuthToken(updatedRefreshTokenContent, currentTimestamp, updatedRefreshTokenLifetime)
+      override def getAuthResponse(
+        accessTokenContent: AccessTokenContent,
+        refreshTokenContent: RefreshTokenContent,
+        currentTimestamp: Long
+      ): Option[AuthResponse] = {
 
-      AuthResponse(newAccessToken, newRefreshToken, currentTimestamp + updatedAccessTokenLifetime)
-    }
+        import authConfig._
 
-    refreshTokenContent.optRestOfUserSession match {
-      case optRestOfUserSession if optRestOfUserSession.exists(_ > 0) || optRestOfUserSession.isEmpty =>
-        Some(buildAuthResponse)
-      case _ =>
-        None
-    }
-  }
+        def buildAuthResponse: AuthResponse = {
+          val updatedAccessTokenLifetime = refreshTokenContent.optRestOfUserSession.map { restOfUserSession =>
+            Seq(restOfUserSession, expirationTimeInSeconds.accessToken).min
+          }.getOrElse(expirationTimeInSeconds.accessToken)
+
+          val updatedRefreshTokenLifetime = refreshTokenContent.optRestOfUserSession.map { restOfUserSession =>
+            Seq(restOfUserSession, expirationTimeInSeconds.refreshToken).min
+          }.getOrElse(expirationTimeInSeconds.refreshToken)
+
+          val updatedRefreshTokenContent = refreshTokenContent.optRestOfUserSession.map { restOfUserSession =>
+            val updatedRestOfUserSession = restOfUserSession - updatedRefreshTokenLifetime
+            refreshTokenContent.copy(optRestOfUserSession = Some(updatedRestOfUserSession))
+          }.getOrElse {
+            val restOfUserSession = expirationTimeInSeconds.userSession - expirationTimeInSeconds.refreshToken
+            refreshTokenContent.copy(optRestOfUserSession = Some(restOfUserSession))
+          }
 
-  def getOptJwtClaims(refreshToken: String): Option[JwtClaim] =
-    Jwt.decode(refreshToken, authConfig.secretKey, Seq(authConfig.hmacAlgorithm)).toOption
+          val newAccessToken = getAuthToken(accessTokenContent, currentTimestamp, updatedAccessTokenLifetime)
+          val newRefreshToken = getAuthToken(updatedRefreshTokenContent, currentTimestamp, updatedRefreshTokenLifetime)
 
-  private def getAuthToken(authContent: AuthContent, currentTimestamp: Long, lifetime: Long): String = {
-    val claims = JwtClaim(
-      content = Json.stringify(Json.toJson(authContent)),
-      expiration = Some(currentTimestamp + lifetime),
-      issuedAt = Some(currentTimestamp)
-    )
-    Jwt.encode(claims, authConfig.secretKey, authConfig.hmacAlgorithm)
-  }
+          AuthResponse(newAccessToken, newRefreshToken, currentTimestamp + updatedAccessTokenLifetime)
+        }
 
+        refreshTokenContent.optRestOfUserSession match {
+          case optRestOfUserSession if optRestOfUserSession.exists(_ > 0) || optRestOfUserSession.isEmpty =>
+            Some(buildAuthResponse)
+          case _ =>
+            None
+        }
```

Comment:
I think this could be changed for more understanding. But it's up to you to decide.
```suggestion
        refreshTokenContent.optRestOfUserSession match {
          case optRestOfUserSession if optRestOfUserSession.exists(_ <= 0) => None
          case _                                                                                                   => Some(buildAuthResponse)
        }
```

Reply:
I prefer not to change code which is not directly involved in the task